pulsar-client-cpp/tests/ClientTest.cc

```diff
@@ -114,3 +116,50 @@ TEST(ClientTest, testConnectTimeout) {
     clientLow.close();
     clientDefault.close();
 }
+
+TEST(ClientTest, testGetNumberOfReferences) {
+    Client client("pulsar://localhost:6650");
+
+    // Producer test
+    uint64_t numberOfProducers = 0;
+    const std::string nonPartitionedTopic = "nonPartitionedTopic";
+    const std::string partitionedTopic = "partitionedTopic";
+    Producer producer;
+    client.createProducer(nonPartitionedTopic, producer);
+    numberOfProducers = 1;
+    ASSERT_EQ(numberOfProducers, client.getNumberOfProducers());
+
+    producer.close();
+    numberOfProducers = 0;
+    ASSERT_EQ(numberOfProducers, client.getNumberOfProducers());
+
+    // PartitionedProducer
+    int res = makePutRequest(
+        "http://localhost:8080/admin/v2/persistent/public/default/" + partitionedTopic + "/partitions", "2");
+    ASSERT_TRUE(res == 204 || res == 409) << "res: " << res;
+
+    client.createProducer(partitionedTopic, producer);
+    numberOfProducers = 2;
+    ASSERT_EQ(numberOfProducers, client.getNumberOfProducers());
+
+    // Consumer test
+    uint64_t numberOfConsumers = 0;
+
+    Consumer consumer;
+    client.subscribe(nonPartitionedTopic, "consumer-1", consumer);
+    numberOfConsumers = 1;
+    ASSERT_EQ(numberOfConsumers, client.getNumberOfConsumers());
+
+    consumer.close();
+    numberOfConsumers = 0;
+    ASSERT_EQ(numberOfConsumers, client.getNumberOfConsumers());
+
+    client.subscribe(partitionedTopic, "consumer-2", consumer);
+    numberOfConsumers = 2;
+    ASSERT_EQ(numberOfConsumers, client.getNumberOfConsumers());
+    client.subscribe(nonPartitionedTopic, "consumer-3", consumer);
+    numberOfConsumers = 3;
+    ASSERT_EQ(numberOfConsumers, client.getNumberOfConsumers());
```

Comment:
It's not good to reuse the `consumer` variable. You can create another two consumers so that you can test if the number of consumers would become 0 after they're closed.